File: ts/packages/ts/src/types/generated.ts
// To parse this data:
//
//   import { Convert, CfguType, Cfgu, Config, ConfigSchemaContentsValue, ConfigSchema, ConfigSet, ConfigStoreQuery, ConfigStoreContentsElement, ConfigStore } from "./file";
//
//   const cfguType = Convert.toCfguType(json);
//   const cfgu = Convert.toCfgu(json);
//   const config = Convert.toConfig(json);
//   const configSchemaContentsValue = Convert.toConfigSchemaContentsValue(json);
//   const configSchemaContents = Convert.toConfigSchemaContents(json);
//   const configSchema = Convert.toConfigSchema(json);
//   const configSet = Convert.toConfigSet(json);
//   const configStoreQuery = Convert.toConfigStoreQuery(json);
//   const configStoreContentsElement = Convert.toConfigStoreContentsElement(json);
//   const configStoreContents = Convert.toConfigStoreContents(json);
//   const configStore = Convert.toConfigStore(json);
//
// These functions will throw an error if the JSON doesn't
// match the expected interface, even if the JSON is valid.

/**
 * A generic declaration of a `Config`, using properties like type, description and
 * constraints.
 * https://configu.com/docs/cfgu/
 */
export interface Cfgu {
    default?:     string;
    depends?:     string[];
    description?: string;
    options?:     string[];
    pattern?:     string;
    required?:    boolean;
    schema?:      { [key: string]: any };
    template?:    string;
    type:         CfguType;
}

export type CfguType = "ARN" | "AWSRegion" | "AZRegion" | "AlibabaRegion" | "Base64" | "Boolean" | "Color" | "ConnectionString" | "Country" | "Currency" | "DateTime" | "DockerImage" | "Domain" | "Email" | "GCPRegion" | "Hex" | "IBMRegion" | "IPv4" | "IPv6" | "JSONSchema" | "Language" | "LatLong" | "Locale" | "MACAddress" | "MD5" | "MIMEType" | "MobilePhone" | "MongoId" | "Number" | "OracleRegion" | "RegEx" | "SHA" | "SemVer" | "String" | "URL" | "UUID";

/**
 * A generic representation of `application configuration` using three properties: `key`,
 * `value`, `set`.
 * https://configu.com/docs/terminology/#config
 */
export interface Config {
    key:   string;
    set:   string;
    value: string;
}

/**
 * A file containing binding records linking each unique `ConfigKey` to its corresponding
 * `Cfgu` declaration.
 * https://configu.com/docs/config-schema/
 */
export interface ConfigSchema {
    contents: { [key: string]: ConfigSchemaContents };
    name:     string;
}

export interface ConfigSchemaContents {
    default?:     string;
    depends?:     string[];
    description?: string;
    options?:     string[];
    pattern?:     string;
    required?:    boolean;
    schema?:      { [key: string]: any };
    template?:    string;
    type:         CfguType;
}

/**
 * A unique, case-sensitive path within a tree-like data structure that groups `Config`s
 * contextually.
 * https://configu.com/docs/config-set/
 */
export interface ConfigSet {
    hierarchy: string[];
    path:      string;
}

export interface ConfigStoreQuery {
    key: string;
    set: string;
}

export interface ConfigStoreContents {
    key:   string;
    set:   string;
    value: string;
}

/**
 * A storage engine interface for `Config`s records.
 * https://configu.com/docs/config-store/
 */
export interface ConfigStore {
    type: string;
}

// Converts JSON strings to/from your types
// and asserts the results of JSON.parse at runtime
export class Convert {
    public static toCfguType(json: string): CfguType {
        return cast(JSON.parse(json), r("CfguType"));
    }

    public static cfguTypeToJson(value: CfguType): string {
        return JSON.stringify(uncast(value, r("CfguType")), null, 2);
    }

    public static toCfgu(json: string): Cfgu {
        return cast(JSON.parse(json), r("Cfgu"));
    }

    public static cfguToJson(value: Cfgu): string {
        return JSON.stringify(uncast(value, r("Cfgu")), null, 2);
    }

    public static toConfig(json: string): Config {
        return cast(JSON.parse(json), r("Config"));
    }

    public static configToJson(value: Config): string {
        return JSON.stringify(uncast(value, r("Config")), null, 2);
    }

    public static toConfigSchemaContentsValue(json: string): ConfigSchemaContents {
        return cast(JSON.parse(json), r("ConfigSchemaContents"));
    }

    public static configSchemaContentsValueToJson(value: ConfigSchemaContents): string {
        return JSON.stringify(uncast(value, r("ConfigSchemaContents")), null, 2);
    }

    public static toConfigSchemaContents(json: string): { [key: string]: ConfigSchemaContents } {
        return cast(JSON.parse(json), m(r("ConfigSchemaContents")));
    }

    public static configSchemaContentsToJson(value: { [key: string]: ConfigSchemaContents }): string {
        return JSON.stringify(uncast(value, m(r("ConfigSchemaContents"))), null, 2);
    }

    public static toConfigSchema(json: string): ConfigSchema {
        return cast(JSON.parse(json), r("ConfigSchema"));
    }

    public static configSchemaToJson(value: ConfigSchema): string {
        return JSON.stringify(uncast(value, r("ConfigSchema")), null, 2);
    }

    public static toConfigSet(json: string): ConfigSet {
        return cast(JSON.parse(json), r("ConfigSet"));
    }

    public static configSetToJson(value: ConfigSet): string {
        return JSON.stringify(uncast(value, r("ConfigSet")), null, 2);
    }

    public static toConfigStoreQuery(json: string): ConfigStoreQuery {
        return cast(JSON.parse(json), r("ConfigStoreQuery"));
    }

    public static configStoreQueryToJson(value: ConfigStoreQuery): string {
        return JSON.stringify(uncast(value, r("ConfigStoreQuery")), null, 2);
    }

    public static toConfigStoreContentsElement(json: string): ConfigStoreContents {
        return cast(JSON.parse(json), r("ConfigStoreContents"));
    }

    public static configStoreContentsElementToJson(value: ConfigStoreContents): string {
        return JSON.stringify(uncast(value, r("ConfigStoreContents")), null, 2);
    }

    public static toConfigStoreContents(json: string): ConfigStoreContents[] {
        return cast(JSON.parse(json), a(r("ConfigStoreContents")));
    }

    public static configStoreContentsToJson(value: ConfigStoreContents[]): string {
        return JSON.stringify(uncast(value, a(r("ConfigStoreContents"))), null, 2);
    }

    public static toConfigStore(json: string): ConfigStore {
        return cast(JSON.parse(json), r("ConfigStore"));
    }

    public static configStoreToJson(value: ConfigStore): string {
        return JSON.stringify(uncast(value, r("ConfigStore")), null, 2);
    }
}

function invalidValue(typ: any, val: any, key: any, parent: any = ''): never {
    const prettyTyp = prettyTypeName(typ);
    const parentText = parent ? ` on ${parent}` : '';
    const keyText = key ? ` for key "${key}"` : '';
    throw Error(`Invalid value${keyText}${parentText}. Expected ${prettyTyp} but got ${JSON.stringify(val)}`);
}

function prettyTypeName(typ: any): string {
    if (Array.isArray(typ)) {
        if (typ.length === 2 && typ[0] === undefined) {
            return `an optional ${prettyTypeName(typ[1])}`;
        } else {
            return `one of [${typ.map(a => { return prettyTypeName(a); }).join(", ")}]`;
        }
    } else if (typeof typ === "object" && typ.literal !== undefined) {
        return typ.literal;
    } else {
        return typeof typ;
    }
}

function jsonToJSProps(typ: any): any {
    if (typ.jsonToJS === undefined) {
        const map: any = {};
        typ.props.forEach((p: any) => map[p.json] = { key: p.js, typ: p.typ });
        typ.jsonToJS = map;
    }
    return typ.jsonToJS;
}

function jsToJSONProps(typ: any): any {
    if (typ.jsToJSON === undefined) {
        const map: any = {};
        typ.props.forEach((p: any) => map[p.js] = { key: p.json, typ: p.typ });
        typ.jsToJSON = map;
    }
    return typ.jsToJSON;
}

function transform(val: any, typ: any, getProps: any, key: any = '', parent: any = ''): any {
    function transformPrimitive(typ: string, val: any): any {
        if (typeof typ === typeof val) return val;
        return invalidValue(typ, val, key, parent);
    }

    function transformUnion(typs: any[], val: any): any {
        // val must validate against one typ in typs
        const l = typs.length;
        for (let i = 0; i < l; i++) {
            const typ = typs[i];
            try {
                return transform(val, typ, getProps);
            } catch (_) {}
        }
        return invalidValue(typs, val, key, parent);
    }

    function transformEnum(cases: string[], val: any): any {
        if (cases.indexOf(val) !== -1) return val;
        return invalidValue(cases.map(a => { return l(a); }), val, key, parent);
    }

    function transformArray(typ: any, val: any): any {
        // val must be an array with no invalid elements
        if (!Array.isArray(val)) return invalidValue(l("array"), val, key, parent);
        return val.map(el => transform(el, typ, getProps));
    }

    function transformDate(val: any): any {
        if (val === null) {
            return null;
        }
        const d = new Date(val);
        if (isNaN(d.valueOf())) {
            return invalidValue(l("Date"), val, key, parent);
        }
        return d;
    }

    function transformObject(props: { [k: string]: any }, additional: any, val: any): any {
        if (val === null || typeof val !== "object" || Array.isArray(val)) {
            return invalidValue(l(ref || "object"), val, key, parent);
        }
        const result: any = {};
        Object.getOwnPropertyNames(props).forEach(key => {
            const prop = props[key];
            const v = Object.prototype.hasOwnProperty.call(val, key) ? val[key] : undefined;
            result[prop.key] = transform(v, prop.typ, getProps, key, ref);
        });
        Object.getOwnPropertyNames(val).forEach(key => {
            if (!Object.prototype.hasOwnProperty.call(props, key)) {
                result[key] = transform(val[key], additional, getProps, key, ref);
            }
        });
        return result;
    }

    if (typ === "any") return val;
    if (typ === null) {
        if (val === null) return val;
        return invalidValue(typ, val, key, parent);
    }
    if (typ === false) return invalidValue(typ, val, key, parent);
    let ref: any = undefined;
    while (typeof typ === "object" && typ.ref !== undefined) {
        ref = typ.ref;
        typ = typeMap[typ.ref];
    }
    if (Array.isArray(typ)) return transformEnum(typ, val);
    if (typeof typ === "object") {
        return typ.hasOwnProperty("unionMembers") ? transformUnion(typ.unionMembers, val)
            : typ.hasOwnProperty("arrayItems")    ? transformArray(typ.arrayItems, val)
            : typ.hasOwnProperty("props")         ? transformObject(getProps(typ), typ.additional, val)
            : invalidValue(typ, val, key, parent);
    }
    // Numbers can be parsed by Date but shouldn't be.
    if (typ === Date && typeof val !== "number") return transformDate(val);
    return transformPrimitive(typ, val);
}

function cast<T>(val: any, typ: any): T {
    return transform(val, typ, jsonToJSProps);
}

function uncast<T>(val: T, typ: any): any {
    return transform(val, typ, jsToJSONProps);
}

function l(typ: any) {
    return { literal: typ };
}

function a(typ: any) {
    return { arrayItems: typ };
}

function u(...typs: any[]) {
    return { unionMembers: typs };
}

function o(props: any[], additional: any) {
    return { props, additional };
}

function m(additional: any) {
    return { props: [], additional };
}

function r(name: string) {
    return { ref: name };
}

const typeMap: any = {
    "Cfgu": o([
        { json: "default", js: "default", typ: u(undefined, "") },
        { json: "depends", js: "depends", typ: u(undefined, a("")) },
        { json: "description", js: "description", typ: u(undefined, "") },
        { json: "options", js: "options", typ: u(undefined, a("")) },
        { json: "pattern", js: "pattern", typ: u(undefined, "") },
        { json: "required", js: "required", typ: u(undefined, true) },
        { json: "schema", js: "schema", typ: u(undefined, m("any")) },
        { json: "template", js: "template", typ: u(undefined, "") },
        { json: "type", js: "type", typ: r("CfguType") },
    ], false),
    "Config": o([
        { json: "key", js: "key", typ: "" },
        { json: "set", js: "set", typ: "" },
        { json: "value", js: "value", typ: "" },
    ], false),
    "ConfigSchema": o([
        { json: "contents", js: "contents", typ: m(r("ConfigSchemaContents")) },
        { json: "name", js: "name", typ: "" },
    ], false),
    "ConfigSchemaContents": o([
        { json: "default", js: "default", typ: u(undefined, "") },
        { json: "depends", js: "depends", typ: u(undefined, a("")) },
        { json: "description", js: "description", typ: u(undefined, "") },
        { json: "options", js: "options", typ: u(undefined, a("")) },
        { json: "pattern", js: "pattern", typ: u(undefined, "") },
        { json: "required", js: "required", typ: u(undefined, true) },
        { json: "schema", js: "schema", typ: u(undefined, m("any")) },
        { json: "template", js: "template", typ: u(undefined, "") },
        { json: "type", js: "type", typ: r("CfguType") },
    ], false),
    "ConfigSet": o([
        { json: "hierarchy", js: "hierarchy", typ: a("") },
        { json: "path", js: "path", typ: "" },
    ], false),
    "ConfigStoreQuery": o([
        { json: "key", js: "key", typ: "" },
        { json: "set", js: "set", typ: "" },
    ], false),
    "ConfigStoreContents": o([
        { json: "key", js: "key", typ: "" },
        { json: "set", js: "set", typ: "" },
        { json: "value", js: "value", typ: "" },
    ], false),
    "ConfigStore": o([
        { json: "type", js: "type", typ: "" },
    ], false),
    "CfguType": [
        "AWSRegion",
        "AZRegion",
        "AlibabaRegion",
        "ARN",
        "Base64",
        "Boolean",
        "Color",
        "ConnectionString",
        "Country",
        "Currency",
        "DateTime",
        "DockerImage",
        "Domain",
        "Email",
        "GCPRegion",
        "Hex",
        "IBMRegion",
        "IPv4",
        "IPv6",
        "JSONSchema",
        "Language",
        "LatLong",
        "Locale",
        "MACAddress",
        "MIMEType",
        "MD5",
        "MobilePhone",
        "MongoId",
        "Number",
        "OracleRegion",
        "RegEx",
        "SHA",
        "SemVer",
        "String",
        "URL",
        "UUID",
    ],
};

File: py/configu/core/generated.py
from enum import Enum
from dataclasses import dataclass
from typing import Optional, List, Dict, Any, TypeVar, Callable, Type, cast


T = TypeVar("T")
EnumT = TypeVar("EnumT", bound=Enum)


def from_str(x: Any) -> str:
    assert isinstance(x, str)
    return x


def from_none(x: Any) -> Any:
    assert x is None
    return x


def from_union(fs, x):
    for f in fs:
        try:
            return f(x)
        except:
            pass
    assert False


def from_list(f: Callable[[Any], T], x: Any) -> List[T]:
    assert isinstance(x, list)
    return [f(y) for y in x]


def from_bool(x: Any) -> bool:
    assert isinstance(x, bool)
    return x


def from_dict(f: Callable[[Any], T], x: Any) -> Dict[str, T]:
    assert isinstance(x, dict)
    return { k: f(v) for (k, v) in x.items() }


def to_enum(c: Type[EnumT], x: Any) -> EnumT:
    assert isinstance(x, c)
    return x.value


def to_class(c: Type[T], x: Any) -> dict:
    assert isinstance(x, c)
    return cast(Any, x).to_dict()


class CfguType(Enum):
    ALIBABA_REGION = "AlibabaRegion"
    ARN = "ARN"
    AWS_REGION = "AWSRegion"
    AZ_REGION = "AZRegion"
    BASE64 = "Base64"
    BOOLEAN = "Boolean"
    COLOR = "Color"
    CONNECTION_STRING = "ConnectionString"
    COUNTRY = "Country"
    CURRENCY = "Currency"
    DATE_TIME = "DateTime"
    DOCKER_IMAGE = "DockerImage"
    DOMAIN = "Domain"
    EMAIL = "Email"
    GCP_REGION = "GCPRegion"
    HEX = "Hex"
    IBM_REGION = "IBMRegion"
    I_PV4 = "IPv4"
    I_PV6 = "IPv6"
    JSON_SCHEMA = "JSONSchema"
    LANGUAGE = "Language"
    LAT_LONG = "LatLong"
    LOCALE = "Locale"
    MAC_ADDRESS = "MACAddress"
    MD5 = "MD5"
    MIME_TYPE = "MIMEType"
    MOBILE_PHONE = "MobilePhone"
    MONGO_ID = "MongoId"
    NUMBER = "Number"
    ORACLE_REGION = "OracleRegion"
    REG_EX = "RegEx"
    SEM_VER = "SemVer"
    SHA = "SHA"
    STRING = "String"
    URL = "URL"
    UUID = "UUID"


@dataclass
class Cfgu:
    """A generic declaration of a `Config`, using properties like type, description and
    constraints.
    https://configu.com/docs/cfgu/
    """
    type: CfguType
    default: Optional[str] = None
    depends: Optional[List[str]] = None
    description: Optional[str] = None
    options: Optional[List[str]] = None
    pattern: Optional[str] = None
    required: Optional[bool] = None
    schema: Optional[Dict[str, Any]] = None
    template: Optional[str] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Cfgu':
        assert isinstance(obj, dict)
        type = CfguType(obj.get("type"))
        default = from_union([from_str, from_none], obj.get("default"))
        depends = from_union([lambda x: from_list(from_str, x), from_none], obj.get("depends"))
        description = from_union([from_str, from_none], obj.get("description"))
        options = from_union([lambda x: from_list(from_str, x), from_none], obj.get("options"))
        pattern = from_union([from_str, from_none], obj.get("pattern"))
        required = from_union([from_bool, from_none], obj.get("required"))
        schema = from_union([lambda x: from_dict(lambda x: x, x), from_none], obj.get("schema"))
        template = from_union([from_str, from_none], obj.get("template"))
        return Cfgu(type, default, depends, description, options, pattern, required, schema, template)

    def to_dict(self) -> dict:
        result: dict = {}
        result["type"] = to_enum(CfguType, self.type)
        if self.default is not None:
            result["default"] = from_union([from_str, from_none], self.default)
        if self.depends is not None:
            result["depends"] = from_union([lambda x: from_list(from_str, x), from_none], self.depends)
        if self.description is not None:
            result["description"] = from_union([from_str, from_none], self.description)
        if self.options is not None:
            result["options"] = from_union([lambda x: from_list(from_str, x), from_none], self.options)
        if self.pattern is not None:
            result["pattern"] = from_union([from_str, from_none], self.pattern)
        if self.required is not None:
            result["required"] = from_union([from_bool, from_none], self.required)
        if self.schema is not None:
            result["schema"] = from_union([lambda x: from_dict(lambda x: x, x), from_none], self.schema)
        if self.template is not None:
            result["template"] = from_union([from_str, from_none], self.template)
        return result


@dataclass
class Config:
    """A generic representation of `application configuration` using three properties: `key`,
    `value`, `set`.
    https://configu.com/docs/terminology/#config
    """
    key: str
    set: str
    value: str

    @staticmethod
    def from_dict(obj: Any) -> 'Config':
        assert isinstance(obj, dict)
        key = from_str(obj.get("key"))
        set = from_str(obj.get("set"))
        value = from_str(obj.get("value"))
        return Config(key, set, value)

    def to_dict(self) -> dict:
        result: dict = {}
        result["key"] = from_str(self.key)
        result["set"] = from_str(self.set)
        result["value"] = from_str(self.value)
        return result


@dataclass
class ConfigSchemaContentsValue:
    type: CfguType
    default: Optional[str] = None
    depends: Optional[List[str]] = None
    description: Optional[str] = None
    options: Optional[List[str]] = None
    pattern: Optional[str] = None
    required: Optional[bool] = None
    schema: Optional[Dict[str, Any]] = None
    template: Optional[str] = None

    @staticmethod
    def from_dict(obj: Any) -> 'ConfigSchemaContentsValue':
        assert isinstance(obj, dict)
        type = CfguType(obj.get("type"))
        default = from_union([from_str, from_none], obj.get("default"))
        depends = from_union([lambda x: from_list(from_str, x), from_none], obj.get("depends"))
        description = from_union([from_str, from_none], obj.get("description"))
        options = from_union([lambda x: from_list(from_str, x), from_none], obj.get("options"))
        pattern = from_union([from_str, from_none], obj.get("pattern"))
        required = from_union([from_bool, from_none], obj.get("required"))
        schema = from_union([lambda x: from_dict(lambda x: x, x), from_none], obj.get("schema"))
        template = from_union([from_str, from_none], obj.get("template"))
        return ConfigSchemaContentsValue(type, default, depends, description, options, pattern, required, schema, template)

    def to_dict(self) -> dict:
        result: dict = {}
        result["type"] = to_enum(CfguType, self.type)
        if self.default is not None:
            result["default"] = from_union([from_str, from_none], self.default)
        if self.depends is not None:
            result["depends"] = from_union([lambda x: from_list(from_str, x), from_none], self.depends)
        if self.description is not None:
            result["description"] = from_union([from_str, from_none], self.description)
        if self.options is not None:
            result["options"] = from_union([lambda x: from_list(from_str, x), from_none], self.options)
        if self.pattern is not None:
            result["pattern"] = from_union([from_str, from_none], self.pattern)
        if self.required is not None:
            result["required"] = from_union([from_bool, from_none], self.required)
        if self.schema is not None:
            result["schema"] = from_union([lambda x: from_dict(lambda x: x, x), from_none], self.schema)
        if self.template is not None:
            result["template"] = from_union([from_str, from_none], self.template)
        return result


@dataclass
class ConfigSchema:
    """A file containing binding records linking each unique `ConfigKey` to its corresponding
    `Cfgu` declaration.
    https://configu.com/docs/config-schema/
    """
    contents: Dict[str, ConfigSchemaContentsValue]
    name: str

    @staticmethod
    def from_dict(obj: Any) -> 'ConfigSchema':
        assert isinstance(obj, dict)
        contents = from_dict(ConfigSchemaContentsValue.from_dict, obj.get("contents"))
        name = from_str(obj.get("name"))
        return ConfigSchema(contents, name)

    def to_dict(self) -> dict:
        result: dict = {}
        result["contents"] = from_dict(lambda x: to_class(ConfigSchemaContentsValue, x), self.contents)
        result["name"] = from_str(self.name)
        return result


@dataclass
class ConfigSet:
    """A unique, case-sensitive path within a tree-like data structure that groups `Config`s
    contextually.
    https://configu.com/docs/config-set/
    """
    hierarchy: List[str]
    path: str

    @staticmethod
    def from_dict(obj: Any) -> 'ConfigSet':
        assert isinstance(obj, dict)
        hierarchy = from_list(from_str, obj.get("hierarchy"))
        path = from_str(obj.get("path"))
        return ConfigSet(hierarchy, path)

    def to_dict(self) -> dict:
        result: dict = {}
        result["hierarchy"] = from_list(from_str, self.hierarchy)
        result["path"] = from_str(self.path)
        return result


@dataclass
class ConfigStoreQuery:
    key: str
    set: str

    @staticmethod
    def from_dict(obj: Any) -> 'ConfigStoreQuery':
        assert isinstance(obj, dict)
        key = from_str(obj.get("key"))
        set = from_str(obj.get("set"))
        return ConfigStoreQuery(key, set)

    def to_dict(self) -> dict:
        result: dict = {}
        result["key"] = from_str(self.key)
        result["set"] = from_str(self.set)
        return result


@dataclass
class ConfigStoreContentsElement:
    key: str
    set: str
    value: str

    @staticmethod
    def from_dict(obj: Any) -> 'ConfigStoreContentsElement':
        assert isinstance(obj, dict)
        key = from_str(obj.get("key"))
        set = from_str(obj.get("set"))
        value = from_str(obj.get("value"))
        return ConfigStoreContentsElement(key, set, value)

    def to_dict(self) -> dict:
        result: dict = {}
        result["key"] = from_str(self.key)
        result["set"] = from_str(self.set)
        result["value"] = from_str(self.value)
        return result


@dataclass
class ConfigStore:
    """A storage engine interface for `Config`s records.
    https://configu.com/docs/config-store/
    """
    type: str

    @staticmethod
    def from_dict(obj: Any) -> 'ConfigStore':
        assert isinstance(obj, dict)
        type = from_str(obj.get("type"))
        return ConfigStore(type)

    def to_dict(self) -> dict:
        result: dict = {}
        result["type"] = from_str(self.type)
        return result


def cfgu_type_from_dict(s: Any) -> CfguType:
    return CfguType(s)


def cfgu_type_to_dict(x: CfguType) -> Any:
    return to_enum(CfguType, x)


def cfgu_from_dict(s: Any) -> Cfgu:
    return Cfgu.from_dict(s)


def cfgu_to_dict(x: Cfgu) -> Any:
    return to_class(Cfgu, x)


def config_from_dict(s: Any) -> Config:
    return Config.from_dict(s)


def config_to_dict(x: Config) -> Any:
    return to_class(Config, x)


def config_schema_contents_value_from_dict(s: Any) -> ConfigSchemaContentsValue:
    return ConfigSchemaContentsValue.from_dict(s)


def config_schema_contents_value_to_dict(x: ConfigSchemaContentsValue) -> Any:
    return to_class(ConfigSchemaContentsValue, x)


def config_schema_contents_from_dict(s: Any) -> Dict[str, ConfigSchemaContentsValue]:
    return from_dict(ConfigSchemaContentsValue.from_dict, s)


def config_schema_contents_to_dict(x: Dict[str, ConfigSchemaContentsValue]) -> Any:
    return from_dict(lambda x: to_class(ConfigSchemaContentsValue, x), x)


def config_schema_from_dict(s: Any) -> ConfigSchema:
    return ConfigSchema.from_dict(s)


def config_schema_to_dict(x: ConfigSchema) -> Any:
    return to_class(ConfigSchema, x)


def config_set_from_dict(s: Any) -> ConfigSet:
    return ConfigSet.from_dict(s)


def config_set_to_dict(x: ConfigSet) -> Any:
    return to_class(ConfigSet, x)


def config_store_query_from_dict(s: Any) -> ConfigStoreQuery:
    return ConfigStoreQuery.from_dict(s)


def config_store_query_to_dict(x: ConfigStoreQuery) -> Any:
    return to_class(ConfigStoreQuery, x)


def config_store_contents_element_from_dict(s: Any) -> ConfigStoreContentsElement:
    return ConfigStoreContentsElement.from_dict(s)


def config_store_contents_element_to_dict(x: ConfigStoreContentsElement) -> Any:
    return to_class(ConfigStoreContentsElement, x)


def config_store_contents_from_dict(s: Any) -> List[ConfigStoreContentsElement]:
    return from_list(ConfigStoreContentsElement.from_dict, s)


def config_store_contents_to_dict(x: List[ConfigStoreContentsElement]) -> Any:
    return from_list(lambda x: to_class(ConfigStoreContentsElement, x), x)


def config_store_from_dict(s: Any) -> ConfigStore:
    return ConfigStore.from_dict(s)


def config_store_to_dict(x: ConfigStore) -> Any:
    return to_class(ConfigStore, x)
